benchexec/container.py

```diff
@@ -716,6 +770,16 @@ def remount_with_additional_flags(mountpoint, fstype, existing_options, mountfla
     libc.mount(None, mountpoint, None, mountflags, None)
 
 
+def escape_overlayfs_parameters(s):
+    """
+    Safely encode a string for being used as a path for overlayfs.
+    In addition to escaping ",", which separates mount options,
+    we need to escape ":", which overlayfs uses to separate multiple lower dirs
+    (cf. https://www.kernel.org/doc/Documentation/filesystems/overlayfs.txt).
+    """
+    return s.replace(b"\\", rb"\\").replace(b":", rb"\:").replace(b",", rb"\,")
```

Comment:
Nice find! I guess this will never happen but why not safeguard against it?

According to the code we need to worry only about `//` in the beginning of the string. So if `s` starts with more than one `/`, lets normalize this to exactly one `/`. And I think we can assume absolute paths here, I would not understand where a relative path would come from. So maybe like

```python
assert s[0] == b'/'
b'/' + s.lstrip(b'/')
```

Reply:
Since the type of `s[0]` is `int`, we should write it like this:

```python
assert s[0] == ord(b"/"), "Path must be absolute"
```

Reply:
```python
    assert s[0] == ord(b"/"), "Path must be absolute"
    normalized_path = b"/" + s.lstrip(b"/")
    return (
        normalized_path.replace(b"\\", rb"\\")
        .replace(b":", rb"\:")
        .replace(b",", rb"\,")
    )
```

Reply:
Nice. Maybe a comment explaining why we normalize with a link to the fuse-overlayfs source? Just make the link a permalink. Or link to https://github.com/sosy-lab/benchexec/pull/1062#discussion_r1742574031.